8254693: Add Panama feature to pass heap segments to native code

JornVernee opened the pull request:
Add the ability to pass heap segments to native code. This requires using `Linker.Option.critical(true)` as a linker option. It has the same limitations as normal critical calls, namely: upcalls into Java are not allowed, and the native function should return relatively quickly. Heap segments are exposed to native code through temporary native addresses that are valid for the duration of the native call.

The motivation for this is supporting existing Java array-based APIs that might have to pass multi-megabyte size arrays to native code, and are current relying on Get-/ReleasePrimitiveArrayCritical from JNI. Where making a copy of the array would be overly prohibitive.

Components of this patch:

- New binding operator `SegmentBase`, which gets the base object of a `MemorySegment`.
- Rename `UnboxAddress` to `SegmentOffset`. Add flag to specify whether processing heap segments should be allowed.
- `CallArranger` impls use new binding operators when `Linker.Option.critical(/* allowHeap= */ true)` is specified.
- `NativeMethodHandle`/`NativeEntryPoint` allow `Object` in their signatures.
- The object/oop + offset is exposed as temporary address to native code.
- Since we stay in the `_thread_in_Java` state, we can safely expose the oops passed to the downcall stub to native code, without needing GCLocker. These oops are valid until we poll for safepoint, which we never do (invoking pure native code).
- Only x64 and AArch64 for now.
- I've refactored `ArgumentShuffle` in the C++ code to no longer rely on callbacks to get the set of source and destination registers (using `CallingConventionClosure`), but instead just rely on 2 equal size arrays with source and destination registers. This allows filtering the input java registers before passing them to `ArgumentShuffle`, which is required to filter out registers holding segment offsets. Replacing placeholder registers is also done as a separate pre-processing step now. See changes in: https://github.com/openjdk/jdk/pull/16201/commits/d2b40f1117d63cc6d74e377bf88cdcf6d15ff866
- I've factored out `DowncallStubGenerator` in the x64 and AArch64 code to use a common `DowncallLinker::StubGenerator`.
- Fallback linker is also supported using JNI's `GetPrimitiveArrayCritical`/`ReleasePrimitiveArrayCritical`

Aside: fixed existing issue with `DowncallLinker` not properly acquiring segments in interpreted mode.

Numbers for the included benchmark on my machine are:

```
Benchmark                     (size)  Mode  Cnt       Score       Error  Units
CriticalCalls.callNotPinned      100  avgt   30      84.818 �     0.729  ns/op
CriticalCalls.callNotPinned    10000  avgt   30    2966.918 �    39.898  ns/op
CriticalCalls.callNotPinned  1000000  avgt   30  952864.052 � 34996.156  ns/op
CriticalCalls.callPinned         100  avgt   30      30.640 �     0.173  ns/op
CriticalCalls.callPinned       10000  avgt   30    2241.403 �    35.473  ns/op
CriticalCalls.callPinned     1000000  avgt   30  221152.247 �  1697.968  ns/op
CriticalCalls.callRecycled       100  avgt   30      40.205 �     0.458  ns/op
CriticalCalls.callRecycled     10000  avgt   30    2845.316 �    13.331  ns/op
CriticalCalls.callRecycled   1000000  avgt   30  287752.178 �  2322.382  ns/op
```

In particular the difference between the `callNotPinned`, which allocates a native segment and copies the heap segment into it, and the `callPinned` which is zero allocation and zero copy, is important. While the allocation can sometimes be avoided (`callRecycled`), sometimes the API's structure prevents allocations from being amortized.

Testing: `jdk_foreign`, testing using a fuzzer.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change requires CSR request [JDK-8318645](https://bugs.openjdk.org/browse/JDK-8318645) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8254693](https://bugs.openjdk.org/browse/JDK-8254693): Add Panama feature to pass heap segments to native code (**Enhancement** - P2)
 * [JDK-8318645](https://bugs.openjdk.org/browse/JDK-8318645): Add Panama feature to pass heap segments to native code (**CSR**)


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [bf850299](https://git.openjdk.org/jdk/pull/16201/files/bf850299ce6d86c15784a224a5905deb134e4530)
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**) ⚠️ Review applies to [fef40cdb](https://git.openjdk.org/jdk/pull/16201/files/fef40cdbff56d2f3ce5aca6a2ef5d06bc72d53ae)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16201/head:pull/16201` \
`$ git checkout pull/16201`

Update a local copy of the PR: \
`$ git checkout pull/16201` \
`$ git pull https://git.openjdk.org/jdk.git pull/16201/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16201`

View PR using the GUI difftool: \
`$ git pr show -t 16201`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16201.diff">https://git.openjdk.org/jdk/pull/16201.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16201#issuecomment-1767613716)

JornVernee commented on test/jdk/java/foreign/critical/TestCritical.java:
```diff
@@ -0,0 +1,177 @@
+                                String fName, FunctionDescriptor fDesc) {}
+
+    @Test(dataProvider = "allowHeapCases")
+    public void testAllowHeap(AllowHeapCase testCase) throws Throwable {
```
Note that this is the only new test. The diff looks bigger because I've renamed the enclosing directory.

JornVernee commented:
@TheRealMDoerr @feilongjiang @sid8606 

I've kept things working as best as possible on PPC, RISC-V, and s390x. But, more changes are needed to actually implement the new feature on those platforms. (See the changes to `downcallLinker_aarch64.cpp`.

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -788,9 +788,23 @@ static StructLayout captureStateLayout() {
          * Using this linker option when linking non-critical functions is likely to have adverse effects,
          * such as loss of performance, or JVM crashes.
+         * <p>
+         * Critical functions can optionally allow access to the Java heap. This allows a client to pass heap
```
```suggestion
         * Critical functions can optionally allow access to the Java heap. This allows clients to pass heap
```

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -788,9 +788,23 @@ static StructLayout captureStateLayout() {
+         * Critical functions can optionally allow access to the Java heap. This allows a client to pass heap
+         * memory segments as addresses, where normally only off-heap memory segments would be allowed. The memory region
+         * inside the Java heap is exposed through a temporary native address that is valid for the duration of the
+         * function call. As such, these temporary addresses, or any addresses derived from them, should not be used
```
The API does not expose temporary addresses - so it is not 100% clear when reading what this para refers to. I suppose you mean a native function that "captures" an object's address and then returns it, so that the client wraps it in a zero-length memory segment? I can't decide on top of my head if this is too cornery or not, even for this javadoc.

JornVernee replied:
I'm not sure what to write to make this clearer. The address that is exposed to the native target function is indeed a temporary address that is constructed from the oop and offset. It is temporary because after the native call, the GC might move the object around, which invalidates the address.

This part is meant to document that native code should not be holding on to the address until after the call completes. This also includes returning the address back to Java. The address would be invalid the moment the function returns.

Would it help if this said: `The memory region inside the Java heap is exposed to the native target function through a temporary native address`?

mcimadamore replied:
I'm considering to leave `As such, these temporary addresses...` out. You already said that the address "is valid for the duration of the call". I'm not sure the subsequent sentence adds much. If you want to make a concrete example of something that should not be done, that would be better. But it seems to me that the client is as in charge as this text suggests - (e.g. if a native lib decides to hold onto an address, the client can't do much about it).

mcimadamore replied:
So, summing up - either we write more and spell what are the things to be on the lookout for, or we leave it out.

JornVernee replied:
Ok, I'll remove the second sentence.

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -788,9 +788,23 @@ static StructLayout captureStateLayout() {
+         * short-lived access to Java heap memory, and copying the relevant data to an off-heap memory segment would be
+         * prohibitive in terms of performance.
+         *
+         * @implNote As a consequence of allowing heap access, the JVM will either lock the garbage collector, or pin the
```
This is a very low-level comment. Which is fine per se. But it doesn't say what does it mean for the developer using this functionality. I think you want to say that GC is impacted in some way or form. If so, please say.

JornVernee replied:
This comment can be removed, as we no longer lock the GC. The same restrictions apply as a 'regular' critical call that does not allow passing heap segments. The call should complete quickly.

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:
```diff
@@ -202,7 +203,7 @@ void interpret(Deque<Object> stack, StoreFunc storeFunc,
                    LoadFunc loadFunc, SegmentAllocator allocator);
 
     private static void checkType(Class<?> type) {
```
For another day: I wonder if we might be better served by more specific "check" routines. E.g. I suppose `Object` is not valid always... and maybe in other cases only certain primitives are ok.

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:
```diff
@@ -268,8 +269,21 @@ static BoxAddress boxAddress(long byteSize) {
+        return SegmentBase.INSTANCE;
+    }
+
+    static SegmentOffset segmentOffsetAllowHeap() {
```
I suppose `heapSegmentOffset` is not good because this binding applies to both native and heap segments, right?

JornVernee replied:
Yes

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:
```diff
@@ -152,8 +152,8 @@ public Bindings getBindings(MethodType mt, FunctionDescriptor cDesc, boolean for
-        BindingCalculator argCalc = forUpcall ? new BoxBindingCalculator(true) : new UnboxBindingCalculator(true, forVariadicFunction);
-        BindingCalculator retCalc = forUpcall ? new UnboxBindingCalculator(false, forVariadicFunction) : new BoxBindingCalculator(false);
+        BindingCalculator argCalc = forUpcall ? new BoxBindingCalculator(true) : new UnboxBindingCalculator(true, forVariadicFunction, options.allowsHeapAccess());
+        BindingCalculator retCalc = forUpcall ? new UnboxBindingCalculator(false, forVariadicFunction, false) : new BoxBindingCalculator(false);
```
asymmetry noted - that said, not much that can be done here - as for an upcall it is not clear what the surrounding context is (e.g. where's the place where unpinning happens)

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/abi/fallback/FallbackLinker.java:
```diff
@@ -180,12 +178,22 @@ private static Object doDowncall(SegmentAllocator returnAllocator, Object[] args
                     acquiredSessions.add(sessionImpl);
-                });
+                    if (invData.allowsHeapAccess() && !ms.isNative()) {
+                        heapBases[i] = ms.heapBase().get();
```
This will fail if the segment is read-only. Use the unsafe variant.

mcimadamore commented on test/micro/org/openjdk/bench/java/lang/foreign/CriticalCalls.java:
```diff
@@ -0,0 +1,113 @@
+    @Benchmark
+    public int callRecycled() throws Throwable {
+        MemorySegment nativeArr = recycler.allocate(JAVA_INT, arr.length);
+        MemorySegment.copy(arr, 0, nativeArr, JAVA_INT, 0, arr.length);
```
Same here - I believe a single call to the correct `allocate` method suffices, but in this case I don't think performance should be skewed too much.

mcimadamore reviewed (commented):
Looks very good. I did a pass of all the Java code, tests and benchmarks. And also looked at the fallback linker (both Java and C). I will leave the hotspot parts to others :-)

JornVernee commented on test/micro/org/openjdk/bench/java/lang/foreign/CriticalCalls.java:
```diff
@@ -0,0 +1,113 @@
+    @Benchmark
+    public int callNotPinned() throws Throwable {
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment nativeArr = arena.allocate(JAVA_INT, arr.length);
```
Good point. I made this benchmark a while ago, before we had the more optimized allocate variants

JornVernee replied:
This had a pretty big impact, actually. Especially on the larger sizes:

```
Benchmark                     (size)  Mode  Cnt       Score       Error  Units
CriticalCalls.callNotPinned      100  avgt   30      84.818 �     0.729  ns/op
CriticalCalls.callNotPinned    10000  avgt   30    2966.918 �    39.898  ns/op
CriticalCalls.callNotPinned  1000000  avgt   30  952864.052 � 34996.156  ns/op
CriticalCalls.callPinned         100  avgt   30      30.640 �     0.173  ns/op
CriticalCalls.callPinned       10000  avgt   30    2241.403 �    35.473  ns/op
CriticalCalls.callPinned     1000000  avgt   30  221152.247 �  1697.968  ns/op
CriticalCalls.callRecycled       100  avgt   30      40.205 �     0.458  ns/op
CriticalCalls.callRecycled     10000  avgt   30    2845.316 �    13.331  ns/op
CriticalCalls.callRecycled   1000000  avgt   30  287752.178 �  2322.382  ns/op
```

mcimadamore replied:
I also notice that the non pinned variant of the `100` benchmark is slow compared to the others. This might be due to try with resources inhibiting scalarization. I suggest to call Arena::close explicitly in that benchmark and repeat the test.

JornVernee replied:
Not sure... The `callNotPinned` variant is meant as a typical use case where the native segment needs to be allocated. I think TWR belongs in that typical use-case.

This is really about measuring the difference between 2 idiomatic code patterns. If non-scalarization is part of one of those patterns, then I think that is something that should be included in the measurement.

JornVernee commented:
Added another benchmark to the patch that xors 2 arrays together using various strategies. These are the results on my machine:

```
Benchmark         (arrayKind)  (sizeKind)  Mode  Cnt   Score    Error  Units
XorTest.xor      JNI_ELEMENTS       SMALL  avgt   30   0.555 �  0.010  ms/op
XorTest.xor      JNI_ELEMENTS      MEDIUM  avgt   30   4.610 �  0.114  ms/op
XorTest.xor      JNI_ELEMENTS       LARGE  avgt   30  53.533 �  2.113  ms/op
XorTest.xor        JNI_REGION       SMALL  avgt   30   0.030 �  0.001  ms/op
XorTest.xor        JNI_REGION      MEDIUM  avgt   30   1.498 �  0.041  ms/op
XorTest.xor        JNI_REGION       LARGE  avgt   30   7.544 �  0.188  ms/op
XorTest.xor      JNI_CRITICAL       SMALL  avgt   30   0.035 �  0.005  ms/op
XorTest.xor      JNI_CRITICAL      MEDIUM  avgt   30   0.496 �  0.003  ms/op
XorTest.xor      JNI_CRITICAL       LARGE  avgt   30   2.521 �  0.035  ms/op
XorTest.xor   FOREIGN_NO_INIT       SMALL  avgt   30   0.030 �  0.001  ms/op
XorTest.xor   FOREIGN_NO_INIT      MEDIUM  avgt   30   1.303 �  0.021  ms/op
XorTest.xor   FOREIGN_NO_INIT       LARGE  avgt   30   7.668 �  0.168  ms/op
XorTest.xor      FOREIGN_INIT       SMALL  avgt   30   0.031 �  0.001  ms/op
XorTest.xor      FOREIGN_INIT      MEDIUM  avgt   30   1.485 �  0.012  ms/op
XorTest.xor      FOREIGN_INIT       LARGE  avgt   30   9.183 �  0.247  ms/op
XorTest.xor  FOREIGN_CRITICAL       SMALL  avgt   30   0.026 �  0.001  ms/op
XorTest.xor  FOREIGN_CRITICAL      MEDIUM  avgt   30   0.501 �  0.002  ms/op
XorTest.xor  FOREIGN_CRITICAL       LARGE  avgt   30   2.578 �  0.023  ms/op
XorTest.xor            UNSAFE       SMALL  avgt   30   0.029 �  0.001  ms/op
XorTest.xor            UNSAFE      MEDIUM  avgt   30   1.300 �  0.013  ms/op
XorTest.xor            UNSAFE       LARGE  avgt   30   7.632 �  0.178  ms/op
```

The important part here is the `FOREIGN_CRITICAL` (the new feature) is on par with `JNI_CRITICAL`.

TheRealMDoerr commented:
PPC64 code is here:
[PPC64_Panama_heap_segments.patch](https://github.com/openjdk/jdk/files/13025334/PPC64_Panama_heap_segments.patch)
All tests are passing on linux PPC64le. One single test case is failing on Big Endian:
test TestLayoutPaths.testBadAlignmentOfRoot(): failure
java.lang.AssertionError: expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertTrue(Assert.java:45)
	at org.testng.Assert.assertTrue(Assert.java:55)
	at TestLayoutPaths.testBadAlignmentOfRoot(TestLayoutPaths.java:157)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:599)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at org.testng.TestRunner.privateRun(TestRunner.java:764)
	at org.testng.TestRunner.run(TestRunner.java:585)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:102)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:333)
	at java.base/java.lang.Thread.run(Thread.java:1570)

JornVernee commented:
> Note: This error is not related to this PR. It was broken by https://github.com/openjdk/jdk/commit/b12c471a990eb8f789410a20084918368c655659 which is incorrect for Big Endian. Should I file a new issue or is that already known?

Please file a new issue.

JornVernee commented:
@TheRealMDoerr Thanks for the patch, I've added it to the PR.

TheRealMDoerr commented:
> @TheRealMDoerr Thanks for the patch, I've added it to the PR.

Thanks for adding it!

I wonder if the native_invoker_size_per_arg thing still works good enough. We may exceed the computed size, now, right?

TheRealMDoerr commented:
> > Note: This error is not related to this PR. It was broken by https://github.com/openjdk/jdk/commit/b12c471a990eb8f789410a20084918368c655659 which is incorrect for Big Endian. Should I file a new issue or is that already known?
> 
> Please file a new issue.

Filed https://bugs.openjdk.org/browse/JDK-8318454.

JornVernee commented:
> I wonder if the native_invoker_size_per_arg thing still works good enough. We may exceed the computed size, now, right?

Good point. I'll have a look at enhancing the test we have for this. 

Intuitively, I think it will be okay. It's true that we generate more code to add the oops and offsets together, but at the same time, we don't have any code to shuffle the offsets.

mcimadamore commented:
/reviewers 2

sid8606 commented:
Add s390x port from here
[S390x_Panama_heap_segments.txt](https://github.com/openjdk/jdk/files/13031418/S390x_Panama_heap_segments.txt)

All tests are passing on linux S390x. One single test case is failing on Big Endian:
test TestLayoutPaths.testBadAlignmentOfRoot(): failure. Similar to PPC Big Endian.

dholmes-ora commented on src/hotspot/share/gc/shared/gcLocker.cpp:
```diff
@@ -136,6 +136,7 @@ void GCLocker::jni_lock(JavaThread* thread) {
     // in the code, but it's too strong; it's possible that the last thread
     // has called `jni_unlock`, but not yet finished the call, e.g. initiating
     // a GCCause::_gc_locker GC.
+    log_debug_jni("Blocked from entering critical section while waiting on GC.");
```
Seems unrelated to current MR - leftover debugging code?

JornVernee replied:
Yes, good catch. Leftover from when this patch used GCLocker (which turned out to be not needed in the end)

TheRealMDoerr commented:
> > I wonder if the native_invoker_size_per_arg thing still works good enough. We may exceed the computed size, now, right?
> 
> Good point. I'll have a look at enhancing the test we have for this.
> 
> Intuitively, I think it will be okay. It's true that we generate more code to add the oops and offsets together, but at the same time, we don't have any code to shuffle the offsets.

Looks like we use 2 input regs per obj, so we reserve 2x native_invoker_size_per_arg. However, the case in which `reg_oop` and `reg_offset` are on stack together with `arg_shuffle` can produce more than this size (depending on platform). If we pass lots of objects on stack, we may exceed the computed size. Correct?

feilongjiang commented:
Here is the patch for risc-v: [riscv_panama_heap_segments.patch](https://github.com/openjdk/jdk/files/13043332/riscv_panama_heap_segments.patch)

 `jdk_foreign` passed with `-XX+VerifyOops -XX:+VerifyStack` with fastdebug build.

RealLucy reviewed (changes requested):
See inline comments for s390 part.
I didn't review all the other code.

RealLucy commented:
@sid8606 You may want to have a look at my s390 comments.

JornVernee commented:
@feilongjiang Thanks, I've added it to the PR

@TheRealMDoerr Note that `reg_offset` is filtered out, and not handled by `arg_shuffle`. So, it becomes the question whether shuffling a register, or adding an offset to an oop takes more bytes. I think most of the `native_invoker_size_per_arg` have soem lenience built in though? (I did do that for x64 and aarch64 at least). So, it might be okay.

I've added an additional test case to the existing test for this, which should stress the new code gen.

sid8606 commented:
@JornVernee please add this patch which addresses @RealLucy comments. 
[0001-Address-Lutz-Schmidt-s-review-comments.txt](https://github.com/openjdk/jdk/files/13046011/0001-Address-Lutz-Schmidt-s-review-comments.txt)

sid8606 commented on src/hotspot/cpu/s390/downcallLinker_s390.cpp:
```diff
@@ -134,28 +79,47 @@ RuntimeStub* DowncallLinker::make_downcall_stub(BasicType* signature,
+      assert(reg_offset.stack_size() == 8, "expected long");
+      Address offset_addr(callerSP, FP_BIAS + reg_offset.offset());
+      __ mem2reg_opt(r_tmp1, offset_addr, true);
+      __ z_agr(reg_oop_reg, r_tmp1);
```
That's right @RealLucy. Thanks for reviewing.

sid8606 commented:
@JornVernee  Please add below patch for addressing of @RealLucy review comments.
[0001-Address-Lutz-Schmidt-s-review-comments.txt](https://github.com/openjdk/jdk/files/13046224/0001-Address-Lutz-Schmidt-s-review-comments.txt)

TheRealMDoerr commented:
> @TheRealMDoerr Note that `reg_offset` is filtered out, and not handled by `arg_shuffle`. So, it becomes the question whether shuffling a register, or adding an offset to an oop takes more bytes. I think most of the `native_invoker_size_per_arg` have some lenience built in though? (I did do that for x64 and aarch64 at least). So, I think it will be okay.

Right.
The filtering happens after the size computation `code_size = native_invoker_code_base_size + (num_args * native_invoker_size_per_arg)`, so we reserve 2 * native_invoker_size_per_arg per heap segment which is 2 * 2 instructions on PPC64.
The maximum actual code size is the size of the 1 * add_offset operation + 1 arg shuffle, which is 4 + 2 instructions on PPC64.
So, we reserve space for 4 instructions, but emit 6 ones.
The reason why it still works is that `_needs_transition` must be false and that saves more space than we exceeded by the oversized argument handling code.
Not a very nice design, but I can live with it.
 
> I've added an additional test case to the existing test for this, which should stress the new code gen. ([dd9e974](https://github.com/openjdk/jdk/commit/dd9e9741de3ca07e6a4cc561002255f98e1e3330))

Thanks for adding it! Would you mind using `limit(84)` which is the maximum? We'd be on the safe side when testing this.

TheRealMDoerr commented:
This is probably the wrong place to ask this, but the sizing topic leads me to another issue: `RuntimeStub::new_runtime_stub` can return `nullptr` when the code cache is full and we would crash when trying to call `nullptr->print_on(&ls)`. Also, what will the Java code do when `downcallStubAddress` is 0 in the `NativeEntryPoint`?
Do you want me to file an issue?

JornVernee commented:
@sid8606 I've added the patch. There were a couple of commented out lines which I've removed.

JornVernee commented:
@TheRealMDoerr 

> So, we reserve space for 4 instructions, but emit 6 ones.

We could bump up the size per arg.

I think I want to get a review for the rest of the hotspot changes first though, to avoid having to go through that process twice, in case changes to the code gen are needed.

> Would you mind using limit(84) which is the maximum?

Why is 84 the maximum? That would result in 170 Java parameters at most, which is well within the 255 limit imposed by the VM spec.

> `RuntimeStub::new_runtime_stub` can return `nullptr` when the code cache is full

As far as I can see we would hit a `fatal` error when the allocation of the runtime stub fails?

https://github.com/openjdk/jdk/blob/4812cabaa489e99481facddce69686a9fee29c44/src/hotspot/share/code/codeBlob.cpp#L425-L430

Note that `alloc_fail_is_fatal` is true by default, and we don't set it when calling `new_runtime_stub`.

Potentially we could bubble up this allocation failure as a Java exception (and do the same for upcall stubs). That would have the benefit of maybe only bringing down a single Java thread, but is at the same time not really a recoverable error, i.e. we can not just disable the linker, like we can for e.g. the JIT compiler. So I'm not sure if the fatal error is really that problematic. We should probably do the same for upcall stubs either way though. I've filed: https://bugs.openjdk.org/browse/JDK-8318586

JornVernee commented:
> Why is 84 the maximum? That would result in 170 Java parameters at most, which is well within the 255 limit imposed by the VM spec.

Ah, right, longs take up 2 slots. So, we get 84*3 = 252 + 4 (target address & return buffer) + 2 receiver method handle + NativeEntryPoint, which gives 258. Ok, I'll find out what the limit is and use that

TheRealMDoerr commented:
> > Why is 84 the maximum? That would result in 170 Java parameters at most, which is well within the 255 limit imposed by the VM spec.
> 
> Ah, right, longs take up 2 slots. So, we get 84*3 = 252 + 4 (target address & return buffer) + receiver method handle + NativeEntryPoint, which gives 258. Ok, I'll find out what the limit is and use that.
> 
> P.S. it's 83

Thanks! I should have mentioned that 84 is the limit for PPC64. 83 should work for all platforms.

RealLucy reviewed (approved):
s390 part looks good now. 
I did not check the shared code nor the other CPU architectures.

mcimadamore commented:
Looking more holistically at the Linker javadoc, there seems to be something missing in that we never say that passing heap segments to downcall is not supported? We have clarified the documentation w.r.t. by-value structs here:

https://github.com/openjdk/panama-foreign/pull/881

But I can't find anything for by-reference structs.

JornVernee commented:
> Looking more holistically at the Linker javadoc, there seems to be something missing in that we never say that passing heap segments to downcall is not supported? We have clarified the documentation w.r.t. by-value structs here:
> 
> [openjdk/panama-foreign#881](https://github.com/openjdk/panama-foreign/pull/881)
> 
> But I can't find anything for by-reference structs.

We added the check as part of: https://github.com/openjdk/panama-foreign/pull/737 I don't see any linker doc update there. I don't think we have anything. We could add a line to `donwcallHandle` that says that the returned handle throws an `IllegalArgumentException` if a heap segment is passed, unless heap segments are explicitly allowed through the `critical(true)` option.

mcimadamore commented:
> We could add a line to `donwcallHandle` that says that the returned handle throws an `IllegalArgumentException` if a heap segment is passed, unless heap segments are explicitly allowed through the `critical(true)` option.

Something like that would be ok, yes.

mcimadamore reviewed (approved):
New javadoc note looks good

JornVernee commented on src/hotspot/share/prims/nativeEntryPoint.cpp:
```diff
@@ -52,10 +52,12 @@ JNI_ENTRY(jlong, NEP_makeDowncallStub(JNIEnv* env, jclass _unused, jobject metho
     oop type_oop = java_lang_invoke_MethodType::ptype(type, i);
-    assert(java_lang_Class::is_primitive(type_oop), "Only primitives expected");
-    BasicType bt = java_lang_Class::primitive_type(type_oop);
+    BasicType bt = java_lang_Class::as_BasicType(type_oop);
```
We can now see `T_OBJECT` here as well, so we just look at the basic type.

JornVernee commented on src/hotspot/share/prims/nativeEntryPoint.cpp:
```diff
@@ -52,10 +52,12 @@ JNI_ENTRY(jlong, NEP_makeDowncallStub(JNIEnv* env, jclass _unused, jobject metho
+    BasicType bt = java_lang_Class::as_BasicType(type_oop);
     basic_type[bt_idx++] = bt;
-    input_regs.push(ForeignGlobals::parse_vmstorage(arg_moves_oop->obj_at(i)));
+    oop reg_oop = arg_moves_oop->obj_at(i);
+    if (reg_oop != nullptr) {
+      input_regs.push(ForeignGlobals::parse_vmstorage(reg_oop));
+    }
```
Some registers are `null`, which indicates that the corresponding value passed to the downcall stub should not be forwarded to the native call, but is instead used directly in the downcall stub (e.g. the offset of an oop).

JornVernee commented on src/hotspot/share/prims/foreignGlobals.cpp:
```diff
@@ -126,62 +126,70 @@ void ArgumentShuffle::print_on(outputStream* os) const {
     // no need for multi-register pairs.
-    //assert(!pair.first()->is_valid() || pair.is_single_reg(), "must be: %s");
-    regs[i] = as_VMStorage(pair.first(), sig_bt[i]);
+    if (signature[i] != T_VOID) {
+      out_regs.push(as_VMStorage(pair.first(), signature[i]));
+    }
```
Since we only care about registers-sized values. It is safe to ignore upper halves of `T_LONG` and `T_DOUBLE` (i.e. `T_VOID`).

JornVernee commented on src/hotspot/cpu/x86/downcallLinker_x86_64.cpp:
```diff
@@ -133,7 +75,41 @@ RuntimeStub* DowncallLinker::make_downcall_stub(BasicType* signature,
+  }
+}
+
+static void runtime_call(MacroAssembler* _masm, address target) {
+  __ vzeroupper();
+  __ mov(r12, rsp); // remember sp
+  __ subptr(rsp, frame::arg_reg_save_area_bytes); // windows
+  __ andptr(rsp, -16); // align stack as required by ABI
+  __ call(RuntimeAddress(target));
+  __ mov(rsp, r12); // restore sp
+  __ reinit_heapbase();
+}
```
This is a minor cleanup to share this code for the three use sites below.

YaSuenag replied:
Question: `r12` does not need to remember?

According to [CallingSequences in OpenJDK Wiki](https://wiki.openjdk.org/display/HotSpot/CallingSequences), `r12` may be reserved for HeapBase if COOP is enabled.  
(`r12` is also used in another places in downcallLinker_x86_64.cpp without restoring...)

JornVernee replied:
You mean `reinit_heapbase` can be removed? I'm not sure whether the caller expects it to be preserved.

Note that `r12` is used in this case to save and restore `rsp`. This is needed since we access data in the frame relative to `rsp`.

merykitty replied:
@YaSuenag `r12` is restored in `reinit_heapbase()` if needed and no, `r12` does not need remembering because it is a constant and can be restored from somewhere else.

TheRealMDoerr replied:
I think your code is fine. Restoring `r12_heapbase` at this point is not bad because `runtime_call` is only for slow paths. I don't think it should be moved.

YaSuenag replied:
Thanks everyone! I understood `r12` is restored in `reinit_heapbase()`. It is necessary. I have no comments.

turbanoff commented on test/jdk/java/foreign/critical/TestStressAllowHeap.java:
```diff
@@ -0,0 +1,111 @@
+            while (true) {
+                try {
+                    doStep(handle, sequence);
+                } catch(Throwable t) {
```
nit
```suggestion
                } catch (Throwable t) {
```

iwanowww commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -776,9 +779,18 @@ static StructLayout captureStateLayout() {
          */
-        static Option critical() {
-            return LinkerOptions.Critical.INSTANCE;
+        static Option critical(boolean allowHeapAccess) {
```
Speaking of public API, I'm surprised to see critical function property conflated with ability to perform on-heap accesses. These aspects look orthogonal to me. Any particular reason not to represent them as 2 separate `Option`s?

Even though it's straightforward to support on-heap accesses during critical function calls, object pinning would support that for non-critical function calls as well, but proposed API doesn't cover it and new API will be required. What's the plan here?

JornVernee replied:
> Even though it's straightforward to support on-heap accesses during critical function calls, object pinning would support that for non-critical function calls as well, but proposed API doesn't cover it and new API will be required. What's the plan here?

The issue is that most GCs don't support object pinning. I don't think we want an API that only works for some GCs? But, if we do, there's a better API that we can have for just pinning support, which is a `MemorySegment::pin(Arena)` method that returns a MemorySegment wrapping the pinned array. That would allow doing multiple native calls with just a single pin operation, and also allows embedding pointers to pinned segments in other data structures.

For the current approach where we make the array accessible for the duration of the native call: without pinning support, other GCs would have to use GCLocker. That means that the native call also has to be relatively short-lived, at which point I figured we might as well drop the thread state transition, since that has the same requirement. I.e. we detect that the call is short-lived, and do the optimization ourselves without the hint from the user (`critical`). This coincidentally also greatly simplifies the implementation. In a prior iteration I did have a separate `allowHeap` `Option` that implied `critical`. But it was suggested to just merge the two together in that case.

mcimadamore replied:
I stand by the current design: a GCLocker-based mechanism (as the current implementation is) needs to have similar restrictions both on-heap access and also removal of state transitions. It's true that a more general notion of pinning is possible, which doesn't necessarily require special support from the linker (because we can turn an heap segment into a native segment by pinning it and _then_ pass that to the linker). But at this point in this support for region-based pinning is not mature enough to justify such an API (and, if we'll ever get to that point, that would not invalidate the critical linker options).

iwanowww reviewed (commented):
hotspot changes look fine.

JornVernee commented on src/hotspot/cpu/aarch64/downcallLinker_aarch64.cpp:
```diff
@@ -193,6 +161,33 @@ void DowncallStubGenerator::generate() {
+  GrowableArray<VMStorage> out_regs = ForeignGlobals::replace_place_holders(_input_registers, locs);
+  ArgumentShuffle arg_shuffle(filtered_java_regs, out_regs, shuffle_reg);
+
+#ifndef PRODUCT
```
This is partly historical. The log output is only intended for debugging, not for end-user eyes. So, I think I originally excluded it as a way of trimming fat from the product build.

Either way, `ArgumentShuffle::print_on` should probably be excluded/included on the same basis.

JornVernee replied:
Either way, this seems like something that should be addressed in a separate patch

JornVernee commented:
/integrate

TheRealMDoerr commented:
One additional comment on the pinning topic: We may even want to pin objects across several downcalls. One downcall could be used to initiate async I/O and other downcalls check the result. The buffer must be stable in the time between them:
"The buffer area being written out must not be accessed during the operation or undefined results may occur.  The memory areas involved must remain valid." https://man7.org/linux/man-pages/man3/aio_write.3.html
(Not sure if we would use on heap memory for that.)

mrserb commented:
Does the usage of this new API can be checked by something like  -Xcheck:jni? Especially a part when the app still doing some upcalls to the JVM from native when the pin is "active".

JornVernee commented:
@mrserb Upcalls are blocked unless a thread is in the `native` thread state [1]. So, if  an upcall happens from a critical function (which doesn't transition to the `native` state), the VM will terminate with a fatal error.

[1]: https://github.com/openjdk/jdk/blob/d5abe49670d81b9c4749ce777ed6bf2886228f2f/src/hotspot/share/prims/upcallLinker.cpp#L79

mrserb commented:
Can I assume it will always cause the fatal error, or it is not specified/UB?

JornVernee commented:
It is not specified, but the current implementation will cause a fatal error.
